src/test/java/org/jenkinsci/plugins/workflow/support/steps/build/RunWrapperTest.java

```diff
@@ -99,4 +106,58 @@
         });
     }
 
+    @Issue("JENKINS-30412")
+    @Test public void getChangeSets() {
+        r.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                sampleRepo1.init();
+                sampleRepo2.init();
+                WorkflowJob p = r.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(
+                    "node {dir('1') {git($/" + sampleRepo1 + "/$)}; dir('2') {git($/" + sampleRepo2 + "/$)}}\n" +
+                    "echo(/changeSets: ${summarize currentBuild}/)\n" +
+                    "@NonCPS def summarize(b) {\n" +
+                    "  b.changeSets.collect {cs ->\n" +
+                    "    /kind=${cs.kind}; entries=/ + cs.collect {entry ->\n" +
+                    "      /${entry.commitId} by ${entry.author.id} ~ ${entry.author.fullName} on ${new Date(entry.timestamp)}: ${entry.msg}: / + entry.affectedFiles.collect {file ->\n" +
+                    "        /${file.editType.name} ${file.path}/\n" +
+                    "      }.join('; ')\n" +
+                    "    }.join(', ')\n" +
+                    "  }.join(' & ')\n" +
+                    "}", true));
+                r.j.assertLogContains("changeSets: ", r.j.assertBuildStatusSuccess(p.scheduleBuild2(0)));
+                sampleRepo1.write("onefile", "stuff");
+                sampleRepo1.git("add", "onefile");
+                sampleRepo1.git("commit", "--message=stuff");
+                assertThat(JenkinsRule.getLog(r.j.assertBuildStatusSuccess(p.scheduleBuild2(0))), containsRegexp(
+                    "changeSets: kind=git; entries=[a-f0-9]{40} by .+ ~ .+ on .+: stuff: add onefile"));
+                sampleRepo1.write("onefile", "more stuff");
+                sampleRepo1.write("anotherfile", "stuff");
+                sampleRepo1.git("add", "onefile", "anotherfile");
+                sampleRepo1.git("commit", "--message=more stuff");
+                sampleRepo1.write("onefile", "amended");
+                sampleRepo1.git("add", "onefile");
+                sampleRepo1.git("commit", "--message=amended");
+                sampleRepo2.write("elsewhere", "stuff");
+                sampleRepo2.git("add", "elsewhere");
+                sampleRepo2.git("commit", "--message=second repo");
+                assertThat(JenkinsRule.getLog(r.j.assertBuildStatusSuccess(p.scheduleBuild2(0))), containsRegexp(
+                    "changeSets: kind=git; entries=[a-f0-9]{40} by .+ ~ .+ on .+: more stuff: (edit onefile; add anotherfile|add anotherfile; edit onefile), " +
+                    "[a-f0-9]{40} by .+ ~ .+ on .+: amended: edit onefile & " +
+                    "kind=git; entries=[a-f0-9]{40} by .+ ~ .+ on .+: second repo: add elsewhere"));
+            }
+        });
+    }
+    // TODO why is this not in Hamcrest to begin with?
```

Comment:
Found it in

```xml
<dependency>
    <groupId>org.hamcrest</groupId>
    <artifactId>java-hamcrest</artifactId>
    <version>2.0.0.0</version>
    <scope>test</scope>
</dependency>
```

but less useful since it does a whole-string match, whereas I want to only look for a substring, so I will stick with this implementation.